control-plane/cni/main.go

```diff
@@ -0,0 +1,223 @@
+	"fmt"
+	"net"
+	"path/filepath"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	current "github.com/containernetworking/cni/pkg/types/100"
+	"github.com/containernetworking/cni/pkg/version"
+	"github.com/hashicorp/go-hclog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+)
+
+const (
+	keyCNIStatus    = "consul.hashicorp.com/cni-status"
+	keyInjectStatus = "consul.hashicorp.com/connect-inject-status"
+	injected        = "injected"
+)
+
+type CNIArgs struct {
+	types.CommonArgs
+	IP                         net.IP
+	K8S_POD_NAME               types.UnmarshallableString
+	K8S_POD_NAMESPACE          types.UnmarshallableString
+	K8S_POD_INFRA_CONTAINER_ID types.UnmarshallableString
+}
+
+// PluginConf is whatever you expect your configuration json to be. This is whatever
+// is passed in on stdin. Your plugin may wish to expose its functionality via
+// runtime args, see CONVENTIONS.md in the CNI spec.
+type PluginConf struct {
+	// This embeds the standard NetConf structure which allows your plugin
+	// to more easily parse standard fields like Name, Type, CNIVersion,
+	// and PrevResult.
+	types.NetConf
+
+	RuntimeConfig *struct {
+		SampleConfig map[string]interface{} `json:"sample_config"`
+	} `json:"runtime_config"`
+
+	Name string `json:"name"`
+	// Type of plugin (consul-cni)
+	Type string `json:"type"`
+	// CNIBinDir is the location of the cni config files on the node. Can bet as a cli flag.
+	CNIBinDir string `json:"cni_bin_dir"`
+	// CNINetDir is the locaion of the cni plugin on the node. Can be set as a cli flag.
+	CNINetDir string `json:"cni_net_dir"`
+	// Multus is if the plugin is a multus plugin. Can be set as a cli flag.
+	Multus bool `json:"multus"`
+	// Kubeconfig file name. Can be set as a cli flag.
+	Kubeconfig string `json:"kubeconfig"`
+	// LogLevl is the logging level. Can be set as a cli flag.
+	LogLevel string `json:"log_level"`
+}
+
+// parseConfig parses the supplied configuration (and prevResult) from stdin.
+func parseConfig(stdin []byte) (*PluginConf, error) {
+	cfg := PluginConf{}
+
+	if err := json.Unmarshal(stdin, &cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse network configuration: %v", err)
+	}
+
+	// Parse previous result. This will parse, validate, and place the
+	// previous result object into conf.PrevResult. If you need to modify
+	// or inspect the PrevResult you will need to convert it to a concrete
+	// versioned Result struct.
+	if err := version.ParsePrevResult(&cfg.NetConf); err != nil {
+		return nil, fmt.Errorf("could not parse prevResult: %v", err)
+	}
+	// End previous result parsing
+
+	// Do any validation here
+	// TODO: Do validation
+	//	if conf.AnotherAwesomeArg == "" {
+	//		return nil, fmt.Errorf("anotherAwesomeArg must be specified")
+	//	}
+
```

Comment:
are we gonna do validation?

Reply:
Yes. I fully expect that we could end up in a scenario where some other plugin contaminates the configuration on us. At the bare minimum I need to make sure that the config is not 'nil'.